### PR TITLE
G2p

### PR DESCRIPTION
**Pull request by dcolligan**

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -9,6 +9,7 @@ cache:
 
 install:
   - pip install pip --upgrade
+  - pip install setuptools==25.1.0
```
Comment (dcolligan) [inline .travis.yml]: Why is this necessary?

File: ga4gh/datamodel/genotype_phenotype.py [hunk 2]
```diff
@@ -0,0 +1,624 @@
...
+
+
+class SimulatedPhenotypeAssociationSet(AbstractPhenotypeAssociationSet):
+    def __init__(self, parentContainer, localId, randomSeed):
+        super(SimulatedPhenotypeAssociationSet, self).__init__(
+            parentContainer, localId)
+
+    # TODO this doesn't make much sense
+    def getAssociations(
+            self, request=None, featureSets=[]):
```
Comment (jeromekelleher) [inline ga4gh/datamodel/genotype_phenotype.py]: Why would request be None here? It would be nice to clarify this a bit.

File: ga4gh/backend.py [hunk 3]
```diff
@@ -700,6 +710,38 @@ def featuresGenerator(self, request):
             request.page_token, request.page_size,
             request.feature_types, parentId, request.name, request.gene_symbol)
 
+    def phenotypesGenerator(self, request):
+        """
+        Returns a generator over the (phenotypes, nextPageToken) pairs
+        defined by the (JSON string) request
+        """
+        # TODO make paging work using SPARQL?
```
Comment (dcolligan) [inline ga4gh/backend.py]: for both `phenotypesGenerator` and `genotypesPhenotypesGenerator` it seems like paging is not implemented efficiently... the entire set of results is requested on each page request, unless I'm missing something (`request.page_token` and `request.page_size` don't appear to be utilized anywhere in the stack)

File: ga4gh/datamodel/genotype_phenotype.py [hunk 4]
```diff
@@ -0,0 +1,624 @@
...
+                return namespace
+
+    def _toGA4GH(self, association):
+        """
+        given an association dict,
+        return a protocol.FeaturePhenotypeAssociation
+        """
+        # TODO: This method needs to broken in several parts. It's
+        # doing too much things
+        fpa = None
```
Comment (jeromekelleher) [inline ga4gh/datamodel/genotype_phenotype.py]: What's this fpa = None for? In general this method looks a bit messy. I think if we had a quick  inline comment describing what each section is for, and remove any dead code and comments, then it would be a big improvement.

File: ga4gh/datamodel/genotype_phenotype.py [hunk 5]
```diff
@@ -0,0 +1,624 @@
...
+
+        # The association dict has the keys: environment, environment
+        # label, evidence, feature label, phenotype and sources. Each
+        # key's value is a dict with the RDF predicates as keys and
+        # subject as values
+
+        # useful
```
Comment (jeromekelleher) [inline ga4gh/datamodel/genotype_phenotype.py]: This is a bit mysterious... Useful for what?

File: ga4gh/datamodel/genotype_phenotype.py [hunk 6]
```diff
@@ -0,0 +1,624 @@
...
+
+class RdfPhenotypeAssociationSet(G2PUtility, AbstractPhenotypeAssociationSet):
+    """
+    An rdf object store.  The cancer genome database
+    [Clinical Genomics Knowledge Base]
+    (http://nif-crawler.neuinfo.org/monarch/ttl/cgd.ttl),
+    published by the Monarch project, was the source of Evidence.
+    """
+    def __init__(self, parentContainer, localId, dataDir):
+        super(RdfPhenotypeAssociationSet, self).__init__(
```
Comment (jeromekelleher) [inline ga4gh/datamodel/genotype_phenotype.py]: Docstring goes before the super() call.

File: ga4gh/datamodel/genotype_phenotype.py [hunk 7]
```diff
@@ -0,0 +1,624 @@
...
+        """
+
+        # initialize graph
+        self._rdfGraph = rdflib.ConjunctiveGraph()
+        # save the path
+        self._dataUrl = dataDir
+
+        try:
+            self._scanDataFiles(dataDir, ['*.ttl', '*.xml'])
+        except AttributeError:
```
Comment (jeromekelleher) [inline ga4gh/datamodel/genotype_phenotype.py]: This looks really dangerous... This will mask any programming errors in the _scanDataFiles method that look for an unknown attribute. Why is this here anyway? The user is much better off if errors happen early and the server fails noisily.

File: ga4gh/frontend.py [hunk 8]
```diff
@@ -451,7 +457,11 @@ def handleFlaskPostRequest(flaskRequest, endpoint):
     Invokes the specified endpoint to generate a response.
     """
     if flaskRequest.method == "POST":
-        return handleHttpPost(flaskRequest, endpoint)
+        try:
```
Comment (jeromekelleher) [inline ga4gh/frontend.py]: Why is this here? This should already be handled by the error handling code.

File: tests/unit/test_data_interface.py [hunk 9]
```diff
@@ -283,6 +284,28 @@ def testSearchReferences(self):
                         repoReferences, references):
                     self._assertEqual(repoReference, reference)
 
+    # TODO
+    @unittest.skip("client and repo don't return same objects")
```
Comment (jeromekelleher) [inline tests/unit/test_data_interface.py]: We should either resolve this or open an issue about it.

File: ga4gh/datamodel/__init__.py [hunk 10]
```diff
@@ -508,6 +529,21 @@ def getParentContainer(self):
         """
         return self._parentContainer
 
+    def _scanDataFiles(self, dataDir, patterns):
```
Comment (dcolligan) [inline ga4gh/datamodel/__init__.py]: Seeing this method added in the diff suggests this pattern has been superseded by another one.  We should probably use a `populateFromFile` and/or `populateFromDirectory` method instead... thoughts, @jeromekelleher ?

**Comment by jeromekelleher** [comment]
Generally looks good to me. There are still some minor issues with dead code and comments lying around, but overall it looks good.

+1 modulo minor comments above.

File: ga4gh/datamodel/genotype_phenotype.py [hunk 11]
```diff
@@ -0,0 +1,624 @@
...
+        return featureType
+
+    def _extractAssociationsDetails(self, associations):
+        """
+        Given a set of results from our search query, return a
+        list of the URIRef for each `detail` (feature,environment,phenotype)
+        """
+        detailedURIRef = []
+        for row in associations.bindings:
+            # empty set [{}]
```
Comment (dcolligan) [inline ga4gh/datamodel/genotype_phenotype.py]: not sure what this comment means

File: tests/unit/test_data_interface.py (hunk 9 above)
Comment (dcolligan) [inline tests/unit/test_data_interface.py]: this is being addressed in #1348

File: ga4gh/datamodel/genotype_phenotype.py [hunk 12]
```diff
@@ -0,0 +1,624 @@
...
+            self._scanDataFiles(dataDir, ['*.ttl', '*.xml'])
+        except AttributeError:
+            pass
+
+        # extract version
+        cgdTTL = rdflib.URIRef("http://data.monarchinitiative.org/ttl/cgd.ttl")
+        versionInfo = rdflib.URIRef(
+            u'http://www.w3.org/2002/07/owl#versionInfo')
+        self._version = None
+        for s, p, o in self._rdfGraph.triples((cgdTTL, versionInfo, None)):
```
Comment (dcolligan) [inline ga4gh/datamodel/genotype_phenotype.py]: we try to avoid single-letter variable names, for clarity

also, in order to not assign names to variables that end up unused in a for loop like this, you can do:

`for _, _, varName in ...`

File: ga4gh/datamodel/genotype_phenotype_featureset.py [hunk 13]
```diff
@@ -0,0 +1,313 @@
...
+            self._scanDataFiles(self._dataUrl, ['*.ttl', '*.xml'])
+        except AttributeError:
+            pass
+
+        # extract version
+        cgdTTL = rdflib.URIRef("http://data.monarchinitiative.org/ttl/cgd.ttl")
+        versionInfo = rdflib.URIRef(
+            u'http://www.w3.org/2002/07/owl#versionInfo')
+        self._version = None
+        for s, p, o in self._rdfGraph.triples((cgdTTL, versionInfo, None)):
```
Comment (dcolligan) [inline ga4gh/datamodel/genotype_phenotype_featureset.py]: same comment about single letter variable names

File: ga4gh/datamodel/genotype_phenotype_featureset.py [hunk 14]
```diff
@@ -0,0 +1,313 @@
...
+
+    def _getFeatureById(self, featureId):
+        """
+        find a feature and return ga4gh representation, use 'native' id as
+        featureId
+        """
+        featureRef = rdflib.URIRef(featureId)
+        featureDetails = self._detailTuples([featureRef])
+        feature = {}
+        for f in featureDetails:
```
Comment (dcolligan) [inline ga4gh/datamodel/genotype_phenotype_featureset.py]: single letter

File: ga4gh/datamodel/genotype_phenotype_featureset.py [hunk 15]
```diff
@@ -0,0 +1,313 @@
...
+        # query to do search
+        query = self._filterSearchFeaturesRequest(
+            referenceName, geneSymbol, name, start, end)
+        featuresResults = self._rdfGraph.query(query)
+        featureIds = set()
+        for row in featuresResults.bindings:
+            featureIds.add(row['feature'].toPython())
+
+        featuresCount = len(featureIds)
+        # else 1 + row number being returned (starting at row 0).
```
Comment (dcolligan) [inline ga4gh/datamodel/genotype_phenotype_featureset.py]: probably some comments that could be removed in the next several lines

File: ga4gh/datamodel/genotype_phenotype_featureset.py [hunk 16]
```diff
@@ -0,0 +1,313 @@
...
+        locationMap[build][chromosome][begin][end] = location["_id"]
+        """
+        # cache of locations
+        self._locationMap = {}
+        locationMap = self._locationMap
+        triples = self._rdfGraph.triples
+        Ref = rdflib.URIRef
+
+        associations = []
+        for s, p, o in triples((None, RDF.type, Ref(ASSOCIATION))):
```
Comment (dcolligan) [inline ga4gh/datamodel/genotype_phenotype_featureset.py]: single letters again... and throughout this method

File: ga4gh/datarepo.py [hunk 17]
```diff
@@ -234,6 +237,14 @@ def printSummary(self):
...
                     featureSet.getId(),
                     sep="\t")
+            print("\tPhenotypeAssociationSets:")
+            for phenotypeAssociationSet in \
+                    dataset.getPhenotypeAssociationSets():
+                print(
+                    "\t", phenotypeAssociationSet.getLocalId(),
+                    phenotypeAssociationSet.getParentContainer().getId(),
+                    sep="\t")
+                # TODO - gabrielle please improve this listing
```
Comment (dcolligan) [inline ga4gh/datarepo.py]: we can keep the comment, but just remove the reference to the person that should do it

File: ga4gh/datarepo.py [hunk 18]
```diff
@@ -565,6 +585,15 @@ def verify(self):
...
                         "\t\t\tRead", i, "annotations from reference",
                         referenceName)
+            for phenotypeAssociationSet \
+                    in dataset.getPhenotypeAssociationSets():
+                print("\t\tVerifying PhenotypeAssociationSet")
+                print(
+                    "\t\t\t", phenotypeAssociationSet.getLocalId(),
+                    phenotypeAssociationSet.getParentContainer().getId(),
+                    sep="\t")
+                # TODO - gabrielle please improve this verification,
```
Comment (dcolligan) [inline ga4gh/datarepo.py]: same comment here

File: ga4gh/datarepo.py [hunk 19]
```diff
@@ -1184,8 +1221,16 @@ def _readFeatureSetTable(self, cursor):
         cursor.execute("SELECT * FROM FeatureSet;")
         for row in cursor:
             dataset = self.getDataset(row[b'datasetId'])
-            featureSet = sequenceAnnotations.Gff3DbFeatureSet(
-                dataset, row[b'name'])
+            # START Load feature set from g2p
```
Comment (dcolligan) [inline ga4gh/datarepo.py]: some comments in the next several lines here can be eliminated

File: tests/end_to_end/test_g2p.py [hunk 20]
```diff
@@ -0,0 +1,530 @@
...
+
+    def testGenotypeSearchFeaturePagingOne(self):
+        """
+        If page size is set to 1 only one association should be returned
+        """
+        request = protocol.SearchFeaturesRequest()
+        request.page_size = 1
+        datasetName, featureSet = self.getCGDDataSetFeatureSet()
+        request.feature_set_id = featureSet.id
+        request.name = \
```
Comment (dcolligan) [inline tests/end_to_end/test_g2p.py]: eliminate backslash

File: tests/end_to_end/test_g2p.py [hunk 21]
```diff
@@ -0,0 +1,530 @@
...
+        self.assertIsNotNone(response.next_page_token)
+
+    def testGenotypeSearchFeaturePagingMore(self):
+        """
+        If page size is not set to more than one association should be returned
+        """
+        request = protocol.SearchFeaturesRequest()
+        datasetName, featureSet = self.getCGDDataSetFeatureSet()
+        request.feature_set_id = featureSet.id
+        request.name = \
```
Comment (dcolligan) [inline tests/end_to_end/test_g2p.py]: eliminate backslash

File: tests/end_to_end/test_g2p.py [hunk 22]
```diff
@@ -0,0 +1,530 @@
...
+
+    def testGenotypeSearchFeaturePagingAll(self):
+        """
+        Loop through all pages
+        """
+        request = protocol.SearchFeaturesRequest()
+        datasetName, featureSet = self.getCGDDataSetFeatureSet()
+        request.feature_set_id = featureSet.id
+        request.page_size = 1
+        request.name = \
```
Comment (dcolligan) [inline tests/end_to_end/test_g2p.py]: seems like a lot of repetition here... this setup could be refactored into a commonly-used method

File: ga4gh/datamodel/__init__.py (hunk 10 above)
Comment (jeromekelleher) [inline ga4gh/datamodel/__init__.py]: If it's a more natural map to the form of the G2P data, then using ``_scanDataFiles`` is fine I think.

File: .travis.yml (hunk 1 above)
Comment (bwalsh) [inline .travis.yml]: travis was complaining see [here](https://travis-ci.org/ga4gh/server/builds/147634111) .

```

Collecting html5lib (from RDFLib==4.2.1->ga4gh==0.3.2.dev130+ng9522f76)
  Downloading html5lib-0.999999999.tar.gz (245kB)
    100% |████████████████████████████████| 245kB 4.9MB/s 
    Complete output from command python setup.py egg_info:
    html5lib requires setuptools version 18.5 or above; please upgrade before installing (you have 12.0.5)
    
    ----------------------------------------
Command "python setup.py egg_info" failed with error code 1 in /tmp/pip-build-2eASFL/html5lib/
The command "pip install dist/ga4gh*.tar.gz" failed and exited with 1 during .
```

File: ga4gh/datamodel/genotype_phenotype.py (hunk 2 above)
Comment (bwalsh) [inline ga4gh/datamodel/genotype_phenotype.py]: @dcolligan This looks like an outdated version? That comment was removed

File: ga4gh/frontend.py (hunk 8 above)
Comment (bwalsh) [inline ga4gh/frontend.py]: @jeromekelleher hi. thanks for your review.  I had this here because underlying exceptions & traces seem to be swallowed up and not logged.  This helped us isolate underlying problems.  I'll remove it.  We may want to revisit in the future if others report similar.

File: .travis.yml (hunk 1 above)
Comment (dcolligan) [inline .travis.yml]: We should add a comment that RDFLib requires it, then

File: ga4gh/datamodel/genotype_phenotype.py (hunk 2 above)
Comment (dcolligan) [inline ga4gh/datamodel/genotype_phenotype.py]: It's being addressed in your PR which hasn't been merged into the g2p branch yet

File: ga4gh/datamodel/__init__.py (hunk 10 above)
Comment (david4096) [inline ga4gh/datamodel/__init__.py]: @dcolligan it would be great if there was one pattern :). The form of the g2p data is just a single `.ttl` file I think.

File: ga4gh/backend.py (hunk 3 above)
Comment (dcolligan) [inline ga4gh/backend.py]: @bwalsh could you comment?

File: ga4gh/datamodel/__init__.py (hunk 10 above)
Comment (bwalsh) [inline ga4gh/datamodel/__init__.py]: the scan has been adjusted to just .ttl

File: ga4gh/backend.py (hunk 3 above)
Comment (bwalsh) [inline ga4gh/backend.py]: At some point in the past, we had looked at implementing paging within the datamodel.  At the time, it seemed that for; conformance with the implementation style with other runners in backend, considering the size of the data (tens, hundred of evidence items not millions) and the discussions about streaming, that we left it as is.

The following tests address paging.  
https://github.com/ga4gh/server/blob/g2p/tests/end_to_end/test_g2p.py#L468
https://github.com/ga4gh/server/blob/g2p/tests/end_to_end/test_g2p.py#L482
https://github.com/ga4gh/server/blob/g2p/tests/end_to_end/test_g2p.py#L495

Suggest considering re-implementation as a performance improvement when needed

**Comment by kozbo** [comment]
+1

File: ga4gh/backend.py (hunk 3 above)
Comment (dcolligan) [inline ga4gh/backend.py]: Ok, that's fine

**Comment by dcolligan** [comment]
Are we waiting on anything to merge this?

**Comment by kozbo** [comment]
@dcolligan We are waiting for the schemas to be ready, but that should be the only thing blocking a merge
